### PR TITLE
Add a Strict monkeypatch to find deprecated behavior

### PR DESCRIPTION
**Pull request by wvanbergen**
Adds a `Strict` monkeypatch that strictly validates the arguments provided to StatsD methods for correct usage. You can enable this in CI of a project by adding `require 'statsd/instrument/strict'`, to fin deprecated behavior that may no longer work in the future.

### CODE REVIEW

**Comment by wvanbergen** [comment]
This is definitely not ready, but some early feedback would be welcome!

File: lib/statsd/instrument.rb [hunk 1]
```diff
@@ -523,7 +523,7 @@ def service_check(
     collect_metric(:_sc, name, status, sample_rate: sample_rate, tags: tags, prefix: prefix, metadata: metadata)
   end
 
-  private
+  protected
```
Comment (dylanahsmith) [inline lib/statsd/instrument.rb]: Why was this change needed?  `protected` doesn't seem like it makes sense on a singleton class.

File: lib/statsd/instrument/strict.rb [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+    # By doing this as part of your QA/CI, you can find where you are still using deprecated patterns,
+    # and fix them before the deprecated behavior is removed in the next major version.
+    #
+    # This monkeypatch is not meant to be used in production.
+    module Strict
+      def increment(key, value = 1, sample_rate: nil, tags: nil, prefix: StatsD.prefix, no_prefix: false)
+        raise ArgumentError, "StatsD.increment does not accept a block" if block_given?
+        raise ArgumentError, "The value argument should be an integer" unless value.is_a?(Integer)
+        raise ArgumentError, "The sample_rate argument should be a number" unless sample_rate.nil? || sample_rate.is_a?(Numeric)
+        raise ArgumentError, "The tags argument should be a hash or an array" unless tags.nil? || tags.is_a?(Hash) || tags.is_a?(Array)
```
Comment (dylanahsmith) [inline lib/statsd/instrument/strict.rb]: We could pull the `sample_rate` and `tags` checks into a common method.

File: lib/statsd/instrument/strict.rb [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+
+      def set(key, value, sample_rate: nil, tags: nil, prefix: StatsD.prefix, no_prefix: false)
+        raise ArgumentError, "The sample_rate argument should be a number" unless sample_rate.nil? || sample_rate.is_a?(Numeric)
+        raise ArgumentError, "The tags argument should be a hash or an array" unless tags.nil? || tags.is_a?(Hash) || tags.is_a?(Array)
+        super
+      end
+
+      def measure(key, value = nil, sample_rate: nil, tags: nil, prefix: StatsD.prefix, no_prefix: false, as_dist: false, &block)
+        if block_given?
+          raise ArgumentError, "The value argument should be set to nil when providing a block" unless value.nil?
```
Comment (dylanahsmith) [inline lib/statsd/instrument/strict.rb]: From the library user's perspective, they should just think that they value shouldn't be given when calling it with a block.

```suggestion
          raise ArgumentError, "A value argument should be given when providing a block" unless value.nil?
```

File: lib/statsd/instrument/version.rb [hunk 4]
```diff
@@ -2,6 +2,6 @@
 
 module StatsD
   module Instrument
-    VERSION = "2.4.0"
+    VERSION = "2.3.9"
```
Comment (dylanahsmith) [inline lib/statsd/instrument/version.rb]: You shouldn't have to lie anymore in Shopify core now that I've resolve the version conflict.

File: test/test_helper.rb [hunk 5]
```diff
@@ -11,4 +11,12 @@
 
 require_relative 'helpers/rubocop_helper'
 
+# Uncomment this line to enable strict mode on the librray itself.
```
Comment (dylanahsmith) [inline test/test_helper.rb]:
```suggestion
# Uncomment this line to enable strict mode on the library itself.
```

You could use an environment variable.  That could also be used to run CI both with and without strict mode by providing the environment variable to one of them.

Comment (wvanbergen) [reply test/test_helper.rb]: Great idea

**Comment by dylanahsmith** [comment]
Should we use sorbet for type checking?  I think we can even have the sorbet rbi file get generated from the documentation to provide that type checking.  We should be able to use the sorbet runtime during testing and not use it for production (using sorbet stubs to ignore any inline annotations in the application).

File: lib/statsd/instrument.rb (hunk 1 above)
Comment (wvanbergen) [reply lib/statsd/instrument.rb]: This was left over from an unsuccessful debugging attempt. It is not needed.

File: test/test_helper.rb [hunk 6]
```diff
@@ -11,4 +11,11 @@
 
 require_relative 'helpers/rubocop_helper'
 
+require 'statsd/instrument/strict' if ENV['STATSD_STRICT_MODE']
+
+def strict_mode_enabled?
```
Comment (casperisfine) [inline test/test_helper.rb]: Not sure i'm fan of having this defined as a main method. Can't it be inside some test helper?

Comment (wvanbergen) [reply test/test_helper.rb]: Fixed

**Comment by wvanbergen** [comment]
So this is working pretty well, and has helped me find many issues in our StatsD usage.

What is still missing is s strict monkeypatch for the metaprgramming methods. I rather add that in a separate PR, so we can review and merge this one.

File: lib/statsd/instrument/strict.rb [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+      def distribution(key, value = UNSPECIFIED, sample_rate: nil, tags: nil,
+        prefix: StatsD.prefix, no_prefix: false, &block)
+
+        check_block_or_numeric_value(value, &block)
+        check_tags_and_sample_rate(sample_rate, tags)
+
+        super
+      end
+
+      protected
```
Comment (dylanahsmith) [inline lib/statsd/instrument/strict.rb]: Looks like this also could be `private`

```suggestion
      private
```
